paddle/framework/executor.cc

```diff
@@ -0,0 +1,177 @@
+      ops_.emplace_back(OpRegistry::CreateOp(op_desc));
+    }
+  }
+}
+
+void GraphView::Initialize(const ProgramDesc* pdesc) {
+  // get a GraphView of ProgramDesc
+}
+
+struct Device {
+  platform::CPUDeviceContext* cpu_device_context;
+#ifndef PADDLE_ONLY_CPU
+  platform::CUDADeviceContext* cuda_device_context;
+#endif
+
+#ifndef PADDLE_ONLY_CPU
+  Device(platform::CPUDeviceContext* cpu, platform::CUDADeviceContext* gpu)
+      : cpu_device_context(cpu), cuda_device_context(gpu) {}
+#else
+  explicit Device(platform::CPUDeviceContext* cpu) : cpu_device_context(cpu) {}
+#endif
+};
+
+class ExecutorImpl : public Executor {
+ public:
+  ExecutorImpl(Scope* scope, const Device* device, const ProgramDesc* pdesc,
+               bool is_linear)
+      : scope_(scope),
+        device_(device),
+        program_desc_(pdesc),
+        view_(ProgramDescView::Create(is_linear)) {}
+
+  virtual ~ExecutorImpl() {
+    if (view_) delete view_;
+  }
+
+  void Run() override;
+
+  void Initialize();
+
+ private:
+  Scope* scope_;
+  const Device* device_;
+  const ProgramDesc* program_desc_;
+  ProgramDescView* view_;
+};
+
+template <typename T, typename... Args>
+std::unique_ptr<T> make_unique(Args&&... args) {
+  return std::unique_ptr<T>(new T(std::forward<Args>(args)...));
+}
+
+platform::CPUDeviceContext* GetCPUDeviceContext(
+    const platform::CPUPlace& place) {
+  static std::unique_ptr<platform::CPUDeviceContext> g_cpu_device_context =
+      make_unique<platform::CPUDeviceContext>(place);
+  return g_cpu_device_context.get();
+}
+
+#ifndef PADDLE_ONLY_CPU
+platform::CUDADeviceContext* GetCUDADeviceContext(
+    const platform::GPUPlace& place) {
+  static std::unique_ptr<platform::CUDADeviceContext> g_cuda_device_context =
+      make_unique<platform::CUDADeviceContext>(place);
+  return g_cuda_device_context.get();
+}
+#endif
+
+Device* GetDevice(const platform::Place& place) {
+  platform::CPUPlace cpu_place;
+#ifndef PADDLE_ONLY_CPU
+  if (platform::is_gpu_place(place)) {
+    platform::GPUPlace gpu_place = boost::get<platform::GPUPlace>(place);
+    static std::unique_ptr<Device> g_device = make_unique<Device>(
+      GetCPUDeviceContext(cpu_place), GetCUDADeviceContext(gpu_place));
+    return g_device.get();
+  } else {
+     static std::unique_ptr<Device> g_device = make_unique<Device>(
+      GetCPUDeviceContext(cpu_place), nullptr);
+    return g_device.get();
+  }
+#else
+  static std::unique_ptr<Device> g_device =
+      make_unique<Device>(GetCPUDeviceContext(cpu_place));
+  return g_device.get();
+#endif
+}
+
+framework::Scope* GetScope() {
+  static std::unique_ptr<framework::Scope> g_scope =
+      make_unique<framework::Scope>();
+  return g_scope.get();
+}
+
+Executor* NewLocalExecutor(const platform::Place& place,
+                           const ProgramDesc& pdesc, bool is_linear) {
+  return new ExecutorImpl(GetScope(), GetDevice(place), &pdesc, is_linear);
+}
+
+void ExecutorImpl::Run() {
```

Comment:
I think one of the most important thing for the executor is Run should be thread-safe (e.g., ok to do concurrent Runs). This is a must for inferencing.

Reply:
That's a good question. We must allow doing concurrent Runs in inference with only one copy of parameters in memory.
I am thinking whom and where to do parameters loading/saving. Our topology can be serialized to `ProgramDesc`, and what will the parameters serialized.